pkg/sidecar/proxy/allowlist.go

```diff
@@ -87,44 +107,79 @@ func NewAllowlistValidator(enabled bool, namespace string, poolName string) (*Al
 		return nil, fmt.Errorf("failed to create Kubernetes dynamic client: %w", err)
 	}
 
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create discovery client: %w", err)
+	}
+
+	var detectedGVR schema.GroupVersionResource
+	for _, gvr := range candidateGVRs {
+		supported, err := isGVRSupported(discoveryClient, gvr)
+		if err != nil {
+			return nil, fmt.Errorf("error checking GVR %s: %w", gvr.String(), err)
+		}
+		if supported {
+			detectedGVR = gvr
+			break
+		}
+	}
+
+	if detectedGVR.Empty() {
+		return nil, fmt.Errorf("no supported InferencePool API found; tried: %v", candidateGVRs)
+	}
+
 	return &AllowlistValidator{
 		enabled:        true,
 		dynamicClient:  dynamicClient,
 		namespace:      namespace,
 		poolName:       poolName,
+		gvr:            detectedGVR,
 		allowedTargets: set.New[string](),
 		podInformers:   make(map[string]cache.SharedInformer),
 		podStopChans:   make(map[string]chan struct{}),
 		stopCh:         make(chan struct{}),
 	}, nil
 }
 
+func isGVRSupported(discoveryClient discovery.DiscoveryInterface, gvr schema.GroupVersionResource) (bool, error) {
+	apiResourceList, err := discoveryClient.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
+	if err != nil {
+		// If the group/version doesn't exist, Kubernetes returns a "NotFound" error
+		if errors.IsNotFound(err) {
+			return false, nil // GroupVersion not supported
+		}
+		return false, fmt.Errorf("failed to discover resources for %s: %w", gvr.GroupVersion(), err)
```

Comment:
nit: as the method is kinda generic doesn't necessarily assume InfPool resource
```suggestion
		return false, fmt.Errorf("failed to discover resources for %s: %w", gvr.String(), err)
```